### PR TITLE
Type checking DAML-LF type synonyms

### PR DESCRIPTION
**Pull request by nickchapman-da**
Extend DAML-LF type-checking to deal with type synonyms.
- checker code here: `compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs`
- driven by tests here: `compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs`

- we changed the AST to ensure type-synonyms are fully applied

TODO: (in a following PR)
- check/disallow cyclic type-synonyms

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 1]
```diff
@@ -145,6 +148,27 @@ kindOf = \case
   TStruct recordType -> checkRecordType recordType $> KStar
   TNat _ -> pure KNat
 
+expandTypeSynonyms :: MonadGamma m => Type -> m Type
+expandTypeSynonyms = expand where
+  expand = \case
+    TVar v -> return $ TVar v
+    TSyn tsyn -> do
+      DefTypeSyn{synType} <- inWorld (lookupTypeSyn tsyn)
```
Comment (cocreature) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: Happy to do this in two steps and not handle type synonyms with parameters as part of this PR but then let’s error out if there are parameters instead of silently doing the wrong thing.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 2]
```diff
@@ -145,6 +148,27 @@ kindOf = \case
   TStruct recordType -> checkRecordType recordType $> KStar
   TNat _ -> pure KNat
 
+expandTypeSynonyms :: MonadGamma m => Type -> m Type
+expandTypeSynonyms = expand where
+  expand = \case
+    TVar v -> return $ TVar v
+    TSyn tsyn -> do
```
Comment (cocreature) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: Given that we have the fully applied type synonym in the protobuf why don’t we preserve this in the Haskell AST and have a constructor of the form `TSynApp TypeSynName [Type]`. That makes it easier to check if it’s correctly applied.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 3]
```diff
@@ -145,6 +148,27 @@ kindOf = \case
...
+      DefTypeSyn{synType} <- inWorld (lookupTypeSyn tsyn)
+      expandTypeSynonyms synType
+    TCon tcon -> return $ TCon tcon
+    TApp tfun targ -> do
+      tfun' <- expand tfun
+      targ' <- expand targ
+      return $ TApp tfun' targ'
+    TBuiltin btype -> return $ TBuiltin btype
+    TForall (v, k) t1 -> do
+      t1' <- expand t1
```
Comment (cocreature) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: Note that `v` can be used as a parameter to a type synonym in `T1` so you might end up having to track variables that are in scope when you add substitution.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs (hunk 1 above)
Comment (nickchapman-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: As discussed, we are adding support for parameterized type synonyms to this PR.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs (hunk 2 above)
Comment (nickchapman-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: Yes. This is much nicer. Done

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs [hunk 4]
```diff
@@ -171,9 +171,9 @@ data Type
   -- | Reference to a type variable.
   = TVar        !TypeVarName
   -- | Reference to a type constructor.
-  | TSyn        !(Qualified TypeSynName)
-  -- | Application of a type function to a type.
   | TCon        !(Qualified TypeConName)
+  -- | Application of a type synonym to its arguments.
```
Comment (cocreature) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs]:
```suggestion
  -- | Fully-applied type synonym
```

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs (hunk 3 above)
Comment (nickchapman-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: I'm pretty sure that because there are no free vars in the body of the type-syn definition, then we should be fine to just use `substitute`

File: language-support/ts/codegen/src/TsCodeGenMain.hs [hunk 5]
```diff
@@ -261,8 +261,8 @@ genType curModName = go
                     ( con' <> "<" <> T.intercalate ", " ts' <> ">"
                     , ser <> "(" <> T.intercalate ", " sers <> ")"
                     )
-        TSyn _ -> error "TODO: genType, type synonym"
         TCon _ -> error "IMPOSSIBLE: lonely type constructor"
+        TSynApp{} -> error "TODO: genType, type synonym"
```
Comment (cocreature) [inline language-support/ts/codegen/src/TsCodeGenMain.hs]: This should be an impossible not a TODO since type synonyms are not serializable.

File: compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs [hunk 6]
```diff
@@ -59,3 +69,166 @@ substitutionTests = testGroup "substitution"
     beta11 = TypeVarName "beta11"
     vBeta1 = TVar beta1
     vBeta11 = TVar beta11
+
+
+typeSynTests :: TestTree
```
Comment (cocreature) [inline compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs]: It would be great if you could add one slightly bigger testcase that tests code a module that is more of the form that we will want to generate, e.g., a single module that defines type synonyms for `Functor` and `Applicative`, instances for one type and a definition that has an `Applicative` constraint. Don’t worry about exactly capturing how the generated code will look like.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 7]
```diff
@@ -554,13 +582,23 @@ checkGroundType ty = do
...
+  exprTypeX <- expandTypeSynonyms exprType
+  typX <- expandTypeSynonyms typ
+  unless (alphaEquiv exprTypeX typX) $
     throwWithContext ETypeMismatch{foundType = exprType, expectedType = typ, expr = Just expr}
   pure exprType
 
 checkExpr :: MonadGamma m => Expr -> Type -> m ()
 checkExpr expr typ = void (checkExpr' expr typ)
 
+-- | Check that a type synonym definition is well-formed.
```
Comment (cocreature) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: This is still missing the check for cyclic type synonyms. Let’s leave that for a separate PR but please add a TODO.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs (hunk 3 above)
Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: :+1:

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs (hunk 4 above)
Comment (nickchapman-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs]: done

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 8]
```diff
@@ -127,8 +127,8 @@ kindOfBuiltin = \case
 kindOf :: MonadGamma m => Type -> m Kind
 kindOf = \case
   TVar v -> lookupTypeVar v
-  TSyn{} -> error "TODO: kindOf, expand type synonym" -- TODO #3616
   TCon tcon -> kindOfDataType <$> inWorld (lookupDataType tcon)
+  TSynApp{} -> return KStar
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: This check will occur in `expandTypeSynonyms`, which is called when an expresion is checked against the type. Is it really possible that this can't happen?

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs (hunk 7 above)
Comment (nickchapman-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: we have this TODO already... In `checkModule` :
```-- TODO #3616: check type synonyms for cycles```

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: You are assuming that there is such an expression. We should check that the type synonym definition is well-kinded even if there is no expression of that type.

File: compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs (hunk 6 above)
Comment (nickchapman-da) [reply compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs]: Not sure exactly what you want to see in this bigger example.
And what are we attempting to test, that isn't already tested?

Comment (cocreature) [reply compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs]: something like the following (in pseudocode)

```
type Functor f = [("fmap", forall a b. (a -> b) -> f a -> f b)]
type Applicative f = [("super", Functor f), ("pure", forall a. a -> f a), ("<*>", forall a b. f (a -> b) -> f b -> f a)]
optFunctor : Functor Optional
optFunctor = …
optApplicative : Applicative Optional
optApplicative = …
f : Applicative f -> a -> f a
f apDict = structProj "pure" apDict
```

We probably do test the individual pieces already but having a test that puts it altogether is helpful both for documentation purposes and to make sure that we didn’t forget anything.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs (hunk 7 above)
Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: :+1:

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: We do check the synonym definition - but that's unrelated to this issue.

This code is computing the kind of a type-expression, which take the form of a syn-application.

You are suggesting that LF may contain a type-expression for which we never check the type of an expression against.

[the remark this answers is not in the document]
Comment (nickchapman-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: I can't imagine how this could happen. But I will repeat the well-formed check on the syn-application anyway.

[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: I’ve pushed a test case that demonstrates this problem. `checkDefTypeSyn` calls `checkType` which in turns just calls `kindOf` so that has to do the check.

[the remark this answers is not in the document]
Comment (nickchapman-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: thanks. nice catch.
fixed now.

File: language-support/ts/codegen/src/TsCodeGenMain.hs (hunk 5 above)
Comment (nickchapman-da) [reply language-support/ts/codegen/src/TsCodeGenMain.hs]: done

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs (hunk 3 above)
Comment (nickchapman-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs]: Turns out we did need to track the variable `v` !

File: compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs (hunk 6 above)
Comment (nickchapman-da) [reply compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs]: bigger example added

Comment (nickchapman-da) [reply compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs]: it found two bugs, so thanks for asking for it

Comment (cocreature) [reply compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs]: Thanks a lot for adding it! It would be great if you could expand it to include the superclass case, i.e., `Applicative` (or `Pointed` or whatever) just to make sure there are no unforeseen issues there either.

Comment (nickchapman-da) [reply compiler/daml-lf-ast/test/DA/Daml/LF/Ast/Tests.hs]: ok, done

**Comment by hurryabit** [comment]
@nickchapman-da This change is not a user-facing change. Thus, there's no need for the changelog entry. I'll remove it from the release notes.
